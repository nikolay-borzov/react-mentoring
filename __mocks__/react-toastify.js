import * as toastify from 'react-toastify'

toastify.toast.error = jest.fn()

export * from 'react-toastify'
