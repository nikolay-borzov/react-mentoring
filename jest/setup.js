import promiseFinally from 'promise.prototype.finally'

import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

// Bring '.finally' to Promise (only for Node 8). Will be available in Node 10
promiseFinally.shim()

Enzyme.configure({ adapter: new Adapter() })

// Make Enzyme functions available in all test files without importing
global.shallow = shallow
global.render = render
global.mount = mount
