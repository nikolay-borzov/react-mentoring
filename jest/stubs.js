export const film = {
  id: 1,
  title: 'Film title',
  poster_path: 'http://via.placeholder.com/500x750',
  overview: 'Film overview',
  release_date: '2018-05-04',
  genres: ['some genre', 'another genre']
}

export const films = [film]
