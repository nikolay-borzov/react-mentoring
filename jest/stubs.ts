import { Film } from '../src/entities/film'

export const film: Film = {
  id: 1,
  title: 'Film title',
  tagline: 'Put that cookie down',
  vote_average: 7.5,
  vote_count: 667,
  release_date: '2018-05-04',
  poster_path: 'https://picsum.photos/g/300/450/?random',
  overview: 'Once upon a time...',
  budget: 15000000,
  revenue: 59735548,
  genres: ['some genre', 'another genre']
}

export const films = [film]

function getRandomImageId() {
  // https://picsum.photos/images
  return Math.floor(Math.random() * Math.floor(1084))
}

export function getRandomFilm(id: number) {
  return {
    ...film,
    id,
    poster_path: film.poster_path.replace(
      'random',
      `image=${getRandomImageId()}`
    )
  }
}
