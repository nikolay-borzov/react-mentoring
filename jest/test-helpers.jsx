import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectly = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectlyShallow = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

/**
 * Creates test closure to check whether component contains another component
 * @param {Function} getComponent Function returning parent component
 * @param {string} componentDisplayName Child component display name
 * @param {Object} params Optional params
 * @param {number} [params.testName=`contains ${componentDisplayName} component`]
 * @param {number} [params.expectedCount=1] Expected number of children components
 * @param {Object} params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent,
  componentDisplayName,
  { testName, expectedCount = 1, expectedProps } = {}
) => {
  it(testName || `contains ${componentDisplayName} component`, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = url => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga, state = {}, arg1 = undefined) => {
  const dispatched = []
  return runSagaOriginal(
    {
      dispatch: action => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
