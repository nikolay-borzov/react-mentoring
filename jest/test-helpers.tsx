import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'
import { AnyAction } from 'redux';

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 */
export const itRendersCorrectly = (
  getComponent: Function,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param getComponent Function returning component to test snapshot
 */
export const itRendersCorrectlyShallow = (
  getComponent: () => React.ReactElement,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

interface ItContainsComponentParams {
  testName?: string
  expectedCount?: number
  expectedProps?: {}
}

/**
 * Creates test closure to check whether component contains another component
 * @param getComponent Function returning parent component
 * @param componentDisplayName Child component display name
 * @param params Optional params
 * @param [params.testName=`contains ${componentDisplayName} component`]
 * @param [params.expectedCount=1] Expected number of children components
 * @param params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent: () => React.ReactElement,
  componentDisplayName: string,
  { testName = `contains ${componentDisplayName} component`, expectedCount = 1, expectedProps }: ItContainsComponentParams = {}
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = (url: string) => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga: any, state = {}, arg1: any = undefined) => {
  const dispatched: AnyAction[] = []
  return runSagaOriginal(
    {
      dispatch: (action: AnyAction) => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
