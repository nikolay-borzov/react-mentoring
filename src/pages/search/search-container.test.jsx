import React from 'react'
import { shallow } from 'enzyme'
import { toast } from 'react-toastify'

import SearchContainerConnected, { SearchContainer } from './search-container'
import configureStore from '../../redux/create'

import { films } from '../../../jest/stubs'

import { setUrl, itRendersCorrectlyShallow } from '../../../jest/test-helpers'

jest.mock('../../services/film-service')

describe('SearchContainer page component', () => {
  const setSearchParamsMock = jest.fn()
  const fetchFilmsMock = jest.fn()
  const historyMock = {
    push: jest.fn()
  }

  const props = {
    match: null,
    history: historyMock,
    searchBy: 'title',
    sortBy: 'release_date',
    films: films,
    filmsError: null,
    foundCount: films.length,
    displayCount: 15,
    isFetching: false,
    setSearchParams: setSearchParamsMock,
    fetchFilms: fetchFilmsMock
  }

  const render = () => {
    const wrapper = shallow(<SearchContainer {...props} />)

    return { wrapper, instance: wrapper.instance() }
  }

  beforeEach(() => {
    setSearchParamsMock.mockClear()
    fetchFilmsMock.mockClear()
    setUrl('/')

    props.match = {
      params: {
        search: ''
      }
    }
    props.films = films
    props.foundCount = films.length

    fetchFilmsMock.mockReturnValue(Promise.resolve())
  })

  afterEach(() => {
    global.IS_SERVER = false
  })

  describe('it renders correctly', () => {
    itRendersCorrectlyShallow(() => {
      const { store } = configureStore()
      return <SearchContainerConnected store={store} />
    }, 'when connected to the store')

    itRendersCorrectlyShallow(
      () => <SearchContainer {...props} />,
      'when films are provided'
    )

    itRendersCorrectlyShallow(() => {
      props.films = []
      props.foundCount = 0

      return <SearchContainer {...props} />
    }, 'when no films found')
  })

  it('generates error for Error Boundary test', () => {
    setUrl('?throwError=1')

    expect(() => render()).toThrowError('Error Boundary test')
  })

  it('loads films on server side', () => {
    global.IS_SERVER = true

    render()

    expect(fetchFilmsMock).toHaveBeenCalled()
  })

  it(`loads films on client side if they haven't been loaded`, () => {
    props.foundCount = 0
    props.match.params.search = 'Comedy'

    render()

    expect(setSearchParamsMock).toHaveBeenCalledWith({
      search: props.match.params.search
    })
    expect(fetchFilmsMock).toHaveBeenCalled()
  })

  /*   it(`saves 'search' from route to store`, () => {
    props.match.params.search = 'Comedy'

    render()

    expect(setSearchParamsMock).toHaveBeenCalledWith({
      search: props.match.params.search
    })
  }) */

  it(`'search' is empty by default`, () => {
    props.foundCount = 0
    props.match.params.search = undefined

    render()

    expect(setSearchParamsMock).toHaveBeenCalledWith({
      search: ''
    })
  })

  it(`updates 'sortBy' on sort change`, () => {
    const sortBy = 'vote_average'

    const { instance } = render()
    instance.onSortByChange(sortBy)

    expect(setSearchParamsMock).toHaveBeenCalledWith({ sortBy })
  })

  it(`updates 'search', 'searchBy' and URL on search change`, () => {
    const params = { search: 'Horror', searchBy: 'genre' }

    const { instance } = render()
    instance.onSearchChange(params)

    expect(setSearchParamsMock).toHaveBeenCalledWith(params)
    expect(historyMock.push).toHaveBeenCalledWith('/search/Horror')
  })

  it(`loads films on 'search' change`, () => {
    const { wrapper, instance } = render()
    // Clear initial call
    fetchFilmsMock.mockClear()

    wrapper.setProps({ match: { params: { search: 'Thriller' } } })
    // https://github.com/airbnb/enzyme/issues/34
    instance.componentDidUpdate(props)

    expect(fetchFilmsMock).toHaveBeenCalled()
  })

  it(`loads films on 'searchBy' change`, () => {
    const { wrapper, instance } = render()
    fetchFilmsMock.mockClear()

    wrapper.setProps({ searchBy: 'genres' })
    instance.componentDidUpdate(props)

    expect(fetchFilmsMock).toHaveBeenCalled()
  })

  it(`loads films on 'sortBy' change`, () => {
    const { wrapper, instance } = render()
    fetchFilmsMock.mockClear()

    wrapper.setProps({ sortBy: 'vote_average' })
    instance.componentDidUpdate(props)

    expect(fetchFilmsMock).toHaveBeenCalled()
  })

  it(`doesn't load films if none of search params has changed`, () => {
    const { wrapper, instance } = render()
    fetchFilmsMock.mockClear()

    wrapper.setProps({ foundCount: 2 })
    instance.componentDidUpdate(props)

    expect(fetchFilmsMock).not.toHaveBeenCalled()
  })

  it('displays an error when unable to load the films', async () => {
    const error = new Error('Films load error')

    const { wrapper, instance } = render()

    wrapper.setProps({ filmsError: error })
    instance.componentDidUpdate(props)

    expect(toast.error).toHaveBeenCalled()
  })
})
