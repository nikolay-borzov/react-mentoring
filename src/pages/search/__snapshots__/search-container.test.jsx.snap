// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchContainer page component it renders correctly when connected to the store 1`] = `
<SearchContainer
  displayCount={0}
  fetchFilms={[Function]}
  films={Array []}
  foundCount={0}
  isFetching={false}
  search=""
  searchBy="title"
  setSearchParams={[Function]}
  sortBy="release_date"
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;

exports[`SearchContainer page component it renders correctly when films are provided 1`] = `
<React.Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Movie Search
    </title>
  </HelmetWrapper>
  <Header>
    <div
      className="padding-controls"
    >
      <SiteName />
      <SearchForm
        onSearchChange={[Function]}
        search=""
        searchBy="title"
      />
    </div>
  </Header>
  <main
    className="content"
  >
    <LoadingBlock
      hideText={false}
      isLoaded={true}
    >
      <SearchResults
        displayCount={15}
        films={
          Array [
            Object {
              "genres": Array [
                "some genre",
                "another genre",
              ],
              "id": 1,
              "overview": "Film overview",
              "poster_path": "http://via.placeholder.com/500x750",
              "release_date": "2018-05-04",
              "title": "Film title",
            },
          ]
        }
        foundCount={1}
        onSortByChange={[Function]}
        sortBy="release_date"
      />
    </LoadingBlock>
  </main>
  <Footer />
</React.Fragment>
`;

exports[`SearchContainer page component it renders correctly when no films found 1`] = `
<React.Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Movie Search
    </title>
  </HelmetWrapper>
  <Header>
    <div
      className="padding-controls"
    >
      <SiteName />
      <SearchForm
        onSearchChange={[Function]}
        search=""
        searchBy="title"
      />
    </div>
  </Header>
  <main
    className="content"
  >
    <LoadingBlock
      hideText={false}
      isLoaded={true}
    >
      <ContentMessage
        className="font-size-header font-bold color-alt"
      >
        No movies found
      </ContentMessage>
    </LoadingBlock>
  </main>
  <Footer />
</React.Fragment>
`;
